### ISSUE TITLE
using a yapdatabase in tests, database isn't deallocating after being set to nil in tearDown

### ISSUE THREAD

**Issue by jonnolen** [yapstudios/YapDatabase#40]
I stand up the database in the test setup and add some data.

Then in the tear down I set the database and the connection to nil... but the database isn't being deallocated, which means [YapDatabaseManager deregisterDatabaseForPath:] isn't being called, so every subsequent test fails because self.database is nil.

Here is a gist with the test and the manual call to YapDatabaseManager deregisterDatabaseForPath:].  

https://gist.github.com/jonnolen/8918335

Is there something that I'm missing here that is preventing the database from being deallocated?

**Comment by cipherCOM** [comment]
We actually have this problem. Our app has a cache in place which is build with YapDatabase. Now during a reset (logout) of the whole app we want to also dismiss the cache. Till now we nilled out the DB and all its connections, waited for its deallocation to be able to delete the whole file and recreate a new DB after this. As this is a async operation including the WAL operations, arbitrary waiting times occur for the user. Is there a better way to do this? Till now we try to go around using ``-removeAllObjectsInAllCollections`` out of performance reasons but seeing some of the higher waiting times we maybe forced to go this route. Is there any better way of performing a complete wipe of the data we're not seeing? We would rather keep the same file path for our cache DB out of simplicity reasons, instead of using randomly generated ones.

**Comment by cipherCOM** [comment]
Investigated further on how to resolve this and found the new API since 2.6.x (we used 2.5.4) with passing a custom connection when registering extensions. This solved the problem of variable wait times for us. Nevertheless we're still searching for a better way of performing a complete wipe instead of waiting for a ``-dealloc`` call of ARC.

**Comment by robbiehanson** [comment]
> We would rather keep the same file path for our cache DB out of simplicity reasons,
> instead of using randomly generated ones.

This is probably the surest route to achieving the performance you seek. Because if you go this route, then any & all async operations won't affect you. You can just tear down your instance variables and you're done.

You can always fetch the file path via ```myYapDatabase.databasePath``` property.

That being said, this is still a valid concern/question:

> we're still searching for a better way of performing a complete wipe instead of waiting
> for a -dealloc call of ARC.

If the delay is purely caused by ```sqlite3_wal_checkpoint_v2```, then there may not be a lot we can do. But understand that I'm saying "caused entirely by that specific sqlite3 function" (meaning its internals, and synchronous disk IO it performs), rather than "caused by something else in YapDatabase land".

In pondering your question, I reviewed the ```asyncCheckpoint:``` method and discovered this little bug:

```objective-c
- (void)asyncCheckpoint:(uint64_t)maxCheckpointableSnapshot
{
	__weak YapDatabase *weakSelf = self;
	
	dispatch_async(checkpointQueue, ^{ @autoreleasepool {
		
		__strong YapDatabase *strongSelf = weakSelf;
		if (strongSelf == nil) return;

		// ...
		// sqlite3_wal_checkpoint_v2 
		// ...

		dispatch_async(strongSelf->snapshotQueue, ^{
			for (YapDatabaseConnectionState *state in strongSelf->connectionStates)
			{
				if (state->longLivedReadTransaction &&
					state->lastTransactionSnapshot == strongSelf->snapshot)
				{
					[state->connection maybeResetLongLivedReadTransaction];
				}
			}
		});
	}});
}
```

In other words, inside the checkpointQueue, if we obtain a strongSelf reference, we maintain it longer than needed by using it within another async operation on the snapshotQueue. If there were a bunch of recently completed readWrite operations, this may have prevented the deallocation of YapDatabase longer than needed.

I just pushed a fix for this. Let me know if that helps at all.

**Comment by cipherCOM** [comment]
Thanks for the quick follow up!

> You can always fetch the file path via ``myYapDatabase.databasePath`` property.

It's not fetching that worries but me but rather the setup. If we loose track of the file path along the way the user will loose her data. So we tried to keep it simple by having one static path.

> If the delay is purely caused by ``sqlite3_wal_checkpoint_v2``, then there may not be a lot we can do.

The main delay previously mentioned was caused by the transient connection created when registering extensions. After finding this and checking back with 2.6.4 we could solve this. Now only the ugly "wait" for ``-dealloc`` remains. Ugly not because of its actual waiting time (only a few ms) but rather the associated block that is called when the database is dealloced. Simply put its an ugly solution not a long waiting period.

> I just pushed a fix for this. Let me know if that helps at all.

As we're not seeing a long waiting period using 2.6.4, I can't say anything about the performance gain. But apart from that, I would propose to not name the internal strong reference ``strongSelf2`` but rather ``strongSelf`` to overshadow the outer one to ensure it can't be used accidentally. Normally we use this pattern ourselves to ensure this.

```
__strong typeof(weakSelf) self = weakSelf;
if (!self) return nil;
```

**Comment by robbiehanson** [comment]
> I would propose to not name the internal strong reference strongSelf2 but rather strongSelf to
> overshadow the outer one to ensure it can't be used accidentally.

Interesting. I assumed that would cause some kind of compiler warning. I'll have to look into this pattern. Thank you.

**Comment by cipherCOM** [comment]
If that is interesting maybe this will look even more intriguing:
```Objective-C
@weakify(self, otherVar);
[self apiWithBlock:^{

    @strongify(self, otherVar);
    if (!self) return;
    ...

}];
```
see [libextobjc/EXTScope.h](https://github.com/jspahrsummers/libextobjc/blob/master/extobjc/EXTScope.h) for various others like ``@onExit``

TIL markdown code highlighting on GitHub from robbiehanson ;)
